### PR TITLE
Opened /api/catalog/books/{book_id}/reviews endpoint for adding a Review

### PR DESCRIPTION
**Pull request by natasatutuspd**

### CODE REVIEW

File: src/main/java/com/productdock/book/ReviewService.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.productdock.book;
+
+import com.productdock.exception.BadRequestException;
+import org.springframework.stereotype.Service;
+
+@Service
+public record ReviewService(ReviewRepository reviewRepository,
```
Comment (milicazivkov85) [inline src/main/java/com/productdock/book/ReviewService.java:7]: missing unit test

**Review by milicazivkov85 — CHANGES_REQUESTED** [review]

File: src/main/java/com/productdock/book/ReviewService.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+                            ReviewMapper reviewMapper) {
+
+    public void saveReview(ReviewDto reviewDto) {
+        var reviewEntity = reviewMapper.toEntity(reviewDto);
+
+        if (reviewRepository.existsById(reviewEntity.getReviewCompositeKey())) {
+            throw new BadRequestException("The user cannot enter more than one comment for a particular book.");
```
Comment (dusanpetrovicpd) [inline src/main/java/com/productdock/book/ReviewService.java:14]: Too generic name, try with something like: BookReviewException

File: src/test/java/com/productdock/book/BookApiTest.java [hunk 3]
```diff
@@ -149,4 +163,65 @@ void getBook_whenTheIdIsExisting() throws Exception {
...
     }
 
+    @Nested
+    class CreateReviewForBook {
+
+        @Test
+        @WithMockUser
+        void createReview_whenReviewIsValid() throws Exception {
+            var reviewDto = defaultReviewDto();
+            var reviewDtoJson = objectMapper.writeValueAsString(reviewDto);
```
Comment (milicazivkov85) [inline src/test/java/com/productdock/book/BookApiTest.java:173]: avoid implementing logic in tests, initialize json from plain string.

File: src/test/java/com/productdock/book/BookApiTest.java [hunk 4]
```diff
@@ -149,4 +163,65 @@ void getBook_whenTheIdIsExisting() throws Exception {
...
+
+            mockMvc.perform(post("/api/catalog/books/1/reviews")
+                            .param("bookId", "1")
```
Comment (milicazivkov85) [inline src/test/java/com/productdock/book/BookApiTest.java:176]: this param is already embedded in post("/api/catalog/books/1/reviews", try without it

File: src/test/java/com/productdock/book/BookApiTest.java [hunk 5]
```diff
@@ -149,4 +163,65 @@ void getBook_whenTheIdIsExisting() throws Exception {
...
+                            })))
+                    .andExpect(status().isOk());
+        }
+
+        @Test
+        @WithMockUser
+        void returnBadRequest_whenReviewAlreadyExists() throws Exception {
+            var reviewDto = defaultReviewDtoBuilder().bookId(1L).build();
+            var review = reviewMapper.toEntity(reviewDto);
+            reviewRepository.save(review);
```
Comment (milicazivkov85) [inline src/test/java/com/productdock/book/BookApiTest.java:191]: In integration tests, entrypoint to the test must be an API call, like it's done in the createReview_whenReviewIsValid() test.

File: src/test/java/com/productdock/book/BookApiTest.java [hunk 6]
```diff
@@ -149,4 +163,65 @@ void getBook_whenTheIdIsExisting() throws Exception {
...
+
+            mockMvc.perform(post("/api/catalog/books/1/reviews")
+                            .param("bookId", "1")
+                            .contentType(MediaType.APPLICATION_JSON)
+                            .content(reviewDtoJson)
+                            .with(jwt().jwt(jwt -> {
+                                jwt.claim("email", "user@mail.com");
+                                jwt.claim("name", "full name");
+                            })))
+                    .andExpect(status().isOk());
```
Comment (milicazivkov85) [inline src/test/java/com/productdock/book/BookApiTest.java:183]: Test will be more reliable if we add another step of verification. Make a call to GET API endpoint, and see if it really returned what has been saved.

File: src/test/java/com/productdock/book/BookApiTest.java [hunk 7]
```diff
@@ -149,4 +163,65 @@ void getBook_whenTheIdIsExisting() throws Exception {
...
+            var reviewDtoJson = objectMapper.writeValueAsString(reviewDto);
+
+            mockMvc.perform(post("/api/catalog/books/1/reviews")
+                            .param("bookId", "1")
+                            .contentType(MediaType.APPLICATION_JSON)
+                            .content(reviewDtoJson)
+                            .with(jwt().jwt(jwt -> {
+                                jwt.claim("email", "user@mail.com");
+                                jwt.claim("name", "full name");
+                            })))
```
Comment (dusanpetrovicpd) [inline src/test/java/com/productdock/book/BookApiTest.java:182]: Make generic method for performing request, and use it for all tests

File: src/test/java/com/productdock/book/ReviewMapperShould.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    @Test
+    void convertReviewDtoToEntity() {
+
+        var reviewDto = ReviewMother.defaultReviewDtoBuilder()
+                .bookId(1L)
+                .recommendation(List.of(MEDIOR, SENIOR))
+                .build();
```
Comment (dusanpetrovicpd) [inline src/test/java/com/productdock/book/ReviewMapperShould.java:48]: This is review dto mother not review mother

**Review by dusanpetrovicpd — CHANGES_REQUESTED** [review]

**Pull request merged**